# Practicumopdrachten Week 2.3 (eerste kans)

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [52]:
import pandas as pd
import sqlite3

import warnings
warnings.simplefilter('ignore')

Zet het bestand go_sales.sqlite in hetzelfde mapje als dit notebook.

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales.sqlite te maken.

In [53]:
gosales_connect = sqlite3.connect("go_sales.sqlite")


Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- target
- returned_item
- return_reason

In [54]:
product_df = pd.read_sql_query("SELECT * FROM product", gosales_connect)
product_type_df = pd.read_sql_query("SELECT * FROM product_type", gosales_connect)
product_line_df = pd.read_sql_query("SELECT * FROM product_line", gosales_connect)
sales_staff_df = pd.read_sql_query("SELECT * FROM sales_staff", gosales_connect)
sales_branch_df = pd.read_sql_query("SELECT * FROM sales_branch", gosales_connect)
retailer_site_df = pd.read_sql_query("SELECT * FROM retailer_site", gosales_connect)
country_df = pd.read_sql_query("SELECT * FROM country", gosales_connect)
order_header_df = pd.read_sql_query("SELECT * FROM order_header", gosales_connect)
order_details_df = pd.read_sql_query("SELECT * FROM order_details", gosales_connect)
target_df = pd.read_sql_query("SELECT * FROM SALES_TARGETData", gosales_connect)
returned_item_df = pd.read_sql_query("SELECT * FROM returned_item", gosales_connect)
return_reason_df = pd.read_sql_query("SELECT * FROM return_reason", gosales_connect)

Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [55]:
product_df['PRODUCTION_COST'] = product_df['PRODUCTION_COST'].astype(float)
product_tussen_50_100 = (product_df['PRODUCTION_COST'] > 50) & (product_df['PRODUCTION_COST'] < 100)
product_df_filter = product_df.loc[(product_tussen_50_100), ['PRODUCT_NAME','PRODUCTION_COST']]
product_df_filter

PRODUCT_NAME  PRODUCTION_COST
13                Blue Steel Putter            60.00
14            Blue Steel Max Putter            81.80
17              Course Pro Golf Bag            80.00
24                  Hibernator Lite            60.00
25                       Hibernator            86.00
28  Hibernator Self - Inflating Mat            78.55
31              Hibernator Camp Cot            65.33
32     Canyon Mule Climber Backpack            62.50
56          Granite Climbing Helmet            52.86
58            Husky Harness Extreme            55.00
69                   Granite Hammer            57.13
71        TrailChef Deluxe Cook Set            85.11
81                        Polar Ice            73.33
83                     Polar Sports            80.00
84                       Polar Wave            60.00
85                    Polar Extreme            72.50
88                     Edge Extreme            80.00
91                        Seeker 35            79.19
92                        Seeker 50            92.58
93           TrailChef Double Flame            75.00
94                   Seeker Extreme            94.12
97                   Glacier Deluxe            56.25
98                      Glacier GPS            78.55

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [59]:
product_margin_20_60 = (product_df['MARGIN'].astype(float) < .2) | (product_df['MARGIN'].astype(float) > .6)
product_df_margin = product_df.loc[(product_margin_20_60), ['PRODUCT_NAME','MARGIN']]
product_df_margin

PRODUCT_NAME MARGIN
29                 Hibernator Pad    .17
32   Canyon Mule Climber Backpack    .17
100             BugShield Natural    .67
101               BugShield Spray    .67
102         BugShield Lotion Lite     .7
103              BugShield Lotion    .63
105             BugShield Extreme    .63

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [61]:
country_franc_as_currency = country_df['CURRENCY_NAME'] == 'francs'
country_franc = country_df.loc[(country_franc_as_currency), 'COUNTRY']
country_franc


0          France
7         Belgium
19    Switzerland
Name: COUNTRY, dtype: object

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [65]:
product_margin_50_introduction = product_df['MARGIN'].astype(float) > .5
product_introduction_date = product_df.loc[(product_margin_50_introduction), ['INTRODUCTION_DATE']].drop_duplicates('INTRODUCTION_DATE')
product_introduction_date

INTRODUCTION_DATE
4          10-12-2019
6          18-12-2019
7          27-12-2019
9           13-1-2020
14         15-12-2019
79         26-10-2016
100         15-2-2011

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 28 rijen)

In [78]:
retailer_adress2_region_known = retailer_site_df['ADDRESS2'].notna() & retailer_site_df['REGION'].notna()
retailer_adress_city = retailer_site_df.loc[(retailer_adress2_region_known), ['ADDRESS1', 'CITY']]
retailer_adress_city


ADDRESS1           CITY
2            4830 Princeton    Albuquerque
5        1700 George Street  San Francisco
13        1488 Orleans Road     Charleston
28      2103 Ashbury Street    Los Angeles
29           333 Ustaff Way      San Diego
..                      ...            ...
367      10423 Kenwood Road     Cincinnati
370   332 South Main Street        Hanover
371     795 South Jefferson    Springfield
376      702 North Meridian  Oklahoma City
377  1210 Somerville Avenue    Chattanooga

[78 rows x 2 columns]

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [89]:
countries_dollar = (country_df['CURRENCY_NAME'] == 'dollars') | (country_df['CURRENCY_NAME'] == 'new dollar')
dollar_countries = country_df.loc[(countries_dollar), 'COUNTRY']
dollar_countries

3            Taiwan
6         Australia
14    United States
15           Canada
Name: COUNTRY, dtype: object

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [98]:
retailer_starts_with_d = (retailer_site_df['POSTAL_ZONE'].str[0] == 'D') & (retailer_site_df['ADDRESS2'].notnull())
retailer_d = retailer_site_df.loc[(retailer_starts_with_d, ['ADDRESS1', 'ADDRESS2', 'CITY'])]
retailer_d

ADDRESS1       ADDRESS2       CITY
205  Röntgenstraße 90  3. Tür rechts  Frankfurt
351   Grubesallee 141       4. Stock    Hamburg

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [103]:
returned_item_df['RETURN_CODE'].count()

706

Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [104]:
retailer_site_df['REGION'].count()

251

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [114]:
import numpy as np
max_margin = np.max(product_df['MARGIN'])
min_margin = np.min(product_df['MARGIN'])
avg_margin = np.mean(product_df['MARGIN'].astype(float))
avg_margin


0.396608695652174

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 9 rijen) 

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

Geef een overzicht met daarin van de verkopers behorend tot de verkoopafdeling uit ‘Amsterdam’ de voornaam en achternaam van de verkoper en zijn laagste en hoogste verkooptarget van 2006. (4 kolommen, 4 rijen) 

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [1]:
from datetime import date
date.today().year

2024

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [2]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 